r/src/compute.cpp

```diff
@@ -574,3 +576,186 @@ SEXP compute__CallFunction(std::string func_name, cpp11::list args, cpp11::list
+          cpp11::sexp output_type_sexp = cpp11::to_r6<arrow::DataType>(output_type);
+          cpp11::writable::list udf_context = {batch_length_sexp, output_type_sexp};
+          udf_context.names() = {"batch_length", "output_type"};
+
+          cpp11::sexp func_result_sexp = state->exec_func_(udf_context, args_sexp);
+
+          if (Rf_inherits(func_result_sexp, "Array")) {
+            auto array = cpp11::as_cpp<std::shared_ptr<arrow::Array>>(func_result_sexp);
+
+            // handle an Array result of the wrong type
+            if (!result->type()->Equals(array->type())) {
+              arrow::Datum out =
+                  ValueOrStop(arrow::compute::Cast(array, result->type()->Copy()));
+              std::shared_ptr<arrow::Array> out_array = out.make_array();
+              array.swap(out_array);
+            }
+
+            // make sure we assign the type that the result is expecting
+            if (result->is_array_data()) {
+              result->value = std::move(array->data());
+            } else if (array->length() == 1) {
+              result->value = ValueOrStop(array->GetScalar(0));
+            } else {
+              cpp11::stop("expected Scalar return value but got Array with length != 1");
+            }
+          } else if (Rf_inherits(func_result_sexp, "Scalar")) {
+            auto scalar = cpp11::as_cpp<std::shared_ptr<arrow::Scalar>>(func_result_sexp);
+
+            // handle a Scalar result of the wrong type
+            if (!result->type()->Equals(scalar->type)) {
+              arrow::Datum out =
+                  ValueOrStop(arrow::compute::Cast(scalar, result->type()->Copy()));
+              std::shared_ptr<arrow::Scalar> out_scalar = out.scalar();
+              scalar.swap(out_scalar);
+            }
+
+            // make sure we assign the type that the result is expecting
+            if (result->is_scalar()) {
+              result->value = std::move(scalar);
+            } else {
+              auto array = ValueOrStop(arrow::MakeArrayFromScalar(
+                  *scalar, span.length, context->memory_pool()));
+              result->value = std::move(array->data());
+            }
+          } else {
+            cpp11::stop("arrow_scalar_function must return an Array or Scalar");
+          }
+        },
+        "execute scalar user-defined function");
+  }
+};
+
+// [[arrow::export]]
+void RegisterScalarUDF(std::string name, cpp11::sexp func_sexp) {
+  cpp11::list in_type_r(func_sexp.attr("in_type"));
+  cpp11::list out_type_r(func_sexp.attr("out_type"));
+  R_xlen_t n_kernels = in_type_r.size();
+
+  if (n_kernels == 0) {
+    cpp11::stop("Can't register user-defined function with zero kernels");
+  }
+
+  // compute the Arity from the list of input kernels
+  std::vector<int64_t> n_args(n_kernels);
+  for (R_xlen_t i = 0; i < n_kernels; i++) {
+    auto in_types = cpp11::as_cpp<std::shared_ptr<arrow::Schema>>(in_type_r[i]);
+    n_args[i] = in_types->num_fields();
```

Comment:
May be? (please test)

```suggestion
    auto in_types = cpp11::as_cpp<std::shared_ptr<arrow::Schema>>(in_type_r[i]);
    n_args[i] = std::move(in_types->num_fields());
```

Reply:
Sorry I didn’t catch that. Yes we can avoid move.